### PR TITLE
Add labels for vm tier

### PR DESCRIPTION
**Pull request by eshen1991**
# Overview
When using [tpu.googleapis.com/accelerator/tensorcore_utilization](https://cloud.google.com/monitoring/api/metrics_gcp) metric, there is no way to break down by VM tiers(i.e. reserved vs spot) as the information is not captured in the existing filtering labels. However the tier information can be injected with VM custom labels, which show up as user labels in the filtering and can be used for grouping by the tiers. The code changes in this PR injects a custom label, vmtier, for TPU VM provisioned via [QRM](https://cloud.google.com/tpu/docs/queued-resources). The custom label supports two values: reserved and spot. 
# Testing
## Unit Testing
```
(axlearn) ➜  gcp git:(vm-tier-label) ✗ pytest tpu_test.py -v
=========================================================== test session starts ===========================================================
platform darwin -- Python 3.9.19, pytest-8.2.2, pluggy-1.5.0 -- /Users/ericshen/miniforge3/envs/axlearn/bin/python3.9
cachedir: .pytest_cache
rootdir: /Users/ericshen/my-git/apple/axlearn
configfile: pyproject.toml
collected 22 items                                                                                                                        

tpu_test.py::TpuUtilsTest::test_create_queued_tpu0 PASSED
tpu_test.py::TpuUtilsTest::test_create_queued_tpu1 PASSED
tpu_test.py::TpuUtilsTest::test_create_queued_tpu2 PASSED
tpu_test.py::TpuUtilsTest::test_create_queued_tpu3 PASSED
tpu_test.py::TpuUtilsTest::test_infer_resources0 PASSED
tpu_test.py::TpuUtilsTest::test_infer_resources1 PASSED
tpu_test.py::TpuUtilsTest::test_infer_resources2 PASSED
tpu_test.py::TpuUtilsTest::test_infer_resources3 PASSED
tpu_test.py::TpuUtilsTest::test_infer_tpu_type0 PASSED
tpu_test.py::TpuUtilsTest::test_infer_tpu_type1 PASSED
tpu_test.py::TpuUtilsTest::test_infer_tpu_type2 PASSED
tpu_test.py::TpuUtilsTest::test_infer_tpu_type3 PASSED
tpu_test.py::TpuUtilsTest::test_infer_utils0 PASSED
tpu_test.py::TpuUtilsTest::test_infer_utils1 PASSED
tpu_test.py::TpuUtilsTest::test_infer_utils2 PASSED
tpu_test.py::TpuUtilsTest::test_queued_resource_info_table PASSED
tpu_test.py::TpuUtilsTest::test_tpu_body_reserved_param0 PASSED
tpu_test.py::TpuUtilsTest::test_tpu_body_reserved_param1 PASSED
tpu_test.py::TpuUtilsTest::test_tpu_body_reserved_param2 PASSED
tpu_test.py::TpuUtilsTest::test_tpu_body_reserved_param3 PASSED
tpu_test.py::TpuUtilsTest::test_tpu_info_table PASSED
tpu_test.py::TpuUtilsTest::test_unknown_tpu_version PASSED

------------------- generated xml file: /Users/ericshen/my-git/apple/axlearn/axlearn/cloud/gcp/test-results/testing.xml -------------------
=========================================================== 22 passed in 0.28s ============================================================

```

## e2e Testing (manual)

1. run ```BASTION_TIER=1 axlearn gcp tpu start --name=$USER-test --tpu_type=v4-8 -- python3 -c "'import jax; print(jax.devices())'"``` to start axlearn job on TPU that is provisioned through [queue resource](https://cloud.google.com/tpu/docs/queued-resources).
2. Once the queued resource state is active and run ```gcloud alpha compute tpus tpu-vm describe ericshen-test --zone us-central2-b``` to check the TPU VM info which should have lines as below:
```
labels:
  vmtier: spot 
schedulingConfig:
  spot: true
```

### CODE REVIEW

File: axlearn/cloud/gcp/tpu.py [hunk 1]
```diff
@@ -567,9 +569,21 @@ def _tpu_body(
...
+    # Create labels for vm tier that can be used to group tpu metrics.
+    # TPUs are created via QRM. In QRM, vm tier can be one of guaranteed, spot, or, bestEffort.
+    # guaranteeded is mapped to label value reserved. bestEffort is implemented to create
+    # spot instance. Hence, both spot and bestEffort are mapped to lable value spot.
+    # See [`_qrm_body()`](#L651) for details on the tiers.
+    if reserved or gcp_settings("reserved_tpu", default=False):
+        vm_tier_labels["vmtier"] = "reserved"
+    else:
+        vm_tier_labels["vmtier"] = "spot"
+
```
Comment (markblee) [inline axlearn/cloud/gcp/tpu.py:582]: Instead of special-casing the vmtier label here, how about letting users supply a dict of `labels` in the TPU creation, similar to `metadata`? That way, the tier logic can be consolidated in the bastion code.

Comment (eshen1991) [reply axlearn/cloud/gcp/tpu.py:582]: @markblee  

> how about letting users supply a dict of labels in the TPU creation, similar to metadata

Are you talking about adding labels as a param in [create_queued_tpu()](https://github.com/apple/axlearn/blob/main/axlearn/cloud/gcp/tpu.py#L293) function  and pass it to the [_tpu_body()](https://github.com/apple/axlearn/blob/main/axlearn/cloud/gcp/tpu.py#L339)? 

> That way, the tier logic can be consolidated in the bastion code

Can you point me to the bastion code?

Comment (markblee) [reply axlearn/cloud/gcp/tpu.py:582]: > Are you talking about adding labels as a param in [create_queued_tpu()](https://github.com/apple/axlearn/blob/main/axlearn/cloud/gcp/tpu.py#L293) function and pass it to the [_tpu_body()](https://github.com/apple/axlearn/blob/main/axlearn/cloud/gcp/tpu.py#L339)?

That's right.

> Can you point me to the bastion code?

This part might be of interest: https://github.com/apple/axlearn/blob/406c3840480422755deb3f9e27806f368f2b7d48/axlearn/cloud/gcp/jobs/tpu_runner.py#L303-L312

Comment (eshen1991) [reply axlearn/cloud/gcp/tpu.py:582]: @markblee please take a look at the [PR 544](https://github.com/apple/axlearn/pull/544),  which has changes that you suggested.

**Review by ruomingp — COMMENTED** [review]
Will defer to @markblee for review.
